### ISSUE TITLE
Incorrectly detects redundant (>>) in where clause 

### ISSUE THREAD

**Issue by cafce25** [ndmitchell/hlint#1207]
When i run HLint  v3.2.7 on
```
main = forever $ putChar 'a'
    where forever a = a >> forever a
```
it tells me i can drop `a >>` and just define `forever a = forever a` but that clearly is not the same IO Action
```
minimal_example.hs:2:23-36: Warning: Redundant >>
Found:
  a >> forever a
Perhaps:
  forever a
```

On the other hand if I define `forever` globally like this:
```
forever a = a >> forever a
```
HLint does not complain

**Comment by googleson78** [comment]
Cause: https://github.com/ndmitchell/hlint/blob/master/data/hlint.yaml#L474

I guess it was assumed that people won't redefine `forever`. But this hint works in general. I don't think it should be removed, since there are a lot of other hints that assume you're not redefining the "standard" thing.

**Comment by cafce25** [comment]
Shouldn't HLint in that case be warning me about redefining forever, maybe suggest removing the whole where clause?
I'm not sure if that is out of scope for HLint?

Totally agree it shouldn't be removed.

**Comment by ndmitchell** [comment]
I don't think HLint saying "depends on the standard definition of forever" is worth it, because almost everything in HLint depends on the standard definitions. I think the reason this isn't firing at the root is that we have a special rule that if you seem to be defining a function `f`, don't suggest replacing it with something that might be the body of `f`. We currently only apply that to top-level definitions, but maybe it should be applied to where clauses too?

**Comment by ndmitchell** [comment]
Yep, I think this is very much a scope issue
